addons/sale_stock/models/sale_order.py

```diff
@@ -125,13 +125,10 @@ def write(self, values):
 
         res = super(SaleOrder, self).write(values)
         if values.get('order_line') and self.state == 'sale':
-            rounding = self.env['decimal.precision'].precision_get('Product Unit of Measure')
             for order in self:
                 to_log = {}
-                for order_line in order.order_line:
-                    if order_line.display_type:
-                        continue
-                    if float_compare(order_line.product_uom_qty, pre_order_line_qty.get(order_line, 0.0), precision_rounding=order_line.product_uom.rounding or rounding) < 0:
+                for order_line in order.order_line.filtered('product_uom'):
```

Comment:
Keep the `continue` inside of the loop instead of use `filtered` to do one loop instead of two (since `filtered` loops on the recordset too) and also, instead of skipping lines with no `product_uom`, I think it's more explicit to precise we want to skip display and downpayment lines
```suggestion
                for order_line:
                    if order_line.display_type or order_line.is_downpayment:
                        continue
```

Reply:
> Keep the `continue` inside of the loop instead of use `filtered` to do one loop instead of two (since `filtered` loops on the recordset too) and also, instead of skipping lines with no `product_uom`, I think it's more explicit to precise we want to skip display and downpayment lines

Doing `filtered` first is actually more efficient as the ORM can optimize it to a single database fetch. If we fetch the field inside of the loop on each iteration, we end up with unneeded database queries that are way costlier than an extra python loop.

Reply:
> and also, instead of skipping lines with no `product_uom`, I think it's more explicit to precise we want to skip display and downpayment lines

I disagree. The field we require inside of the loop is `product_uom.rounding`, neither downpayment nor display make this explicit. This is also more prone to regression, as it's not guaranteed that filtering out `display_type` and `is_downpayment` actually ensures the lines have a `product_uom` value, as the SQL constraint can change again in the future, just as it did between 17.0 and 17.1.

Reply:
> > and also, instead of skipping lines with no `product_uom`, I think it's more explicit to precise we want to skip display and downpayment lines
> 
> I disagree. The field we require inside of the loop is `product_uom.rounding`, neither downpayment nor display make this explicit.

For me, what is explicit is we want to skip downpayment and display lines because functionally, it doesn't make sense to apply following code to them.

> This is also more prone to regression, as it's not guaranteed that filtering out `display_type` and `is_downpayment` actually ensures the lines have a `product_uom` value, as the SQL constraint can change again in the future, just as it did between 17.0 and 17.1.

Indeed, that's kind of my point. If the UoM is required for other cases and we run into a SOL without UoM, I would prefer to get a traceback (making clear there is an issue) instead of skipping the line (making the error silent.)

> Then there's also the fact that fetching two fields is costlier on the ORM than fetching a single field.

That's a good point but is it so significative ? Using `fetch` like you did should mitigate the issue (if there is one)

Reply:
> Indeed, that's kind of my point. If the UoM is required for other cases and we run into a SOL without UoM, I would prefer to get a traceback (making clear there is an issue) instead of skipping the line (making the error silent.)

fair enough, updated it to `if order_line.display_type or order_line.is_downpayment:`